Enhance documentation in regards to modifiers and return (#12129)
ethereum/solidity: docs/ir/ir-breaking-changes.rst

  We have the same behavior for implicit delete, for example when array of structs is shortened.

- Function modifiers are implemented in a slightly different way regarding function parameters and return variables.
  This especially has an effect if the placeholder ``_;`` is evaluated multiple times in a modifier.
  In the old code generator, each function parameter and return variable has a fixed slot on the stack.
  If the function is run multiple times because ``_;`` is used multiple times or used in a loop, then a

>   change to the function parameter's or return variable's value is visible in the next execution of the function.

Reviewer: ```suggestion
  change to the function parameter(s) or return variable(s) value is visible in the next execution of the function.
```

Reviewer: Also `s/execution/invocation/g` may read better.

Reviewer: May be evaluation is even better.

Reviewer 2: Please let's not use `(s)`, everyone will understand that this is equally valid for exactly one parameter or variable.

Reviewer: Please ignore the comment about (s). I misread the sentence.

>   The new code generator implements modifiers using actual functions and passes function parameters on.

Reviewer: ```suggestion
  The new code generator implements modifiers using actual functions and passes function (return) parameters on.
```

Author: I don't think return parameters are "passed" on, rather they are just properly initialized before each call..

>   This means that multiple executions of a function's body will get the same values for the parameters,

Reviewer: ```suggestion
  This means that multiple evaluations of a function's body will get the same values for the parameters,
```

>   and the effect on return variables is that they are reset to their default (zero) value for each

Reviewer: What about non-value type return parameters? E.g., bytes memory, struct memory etc.

Author: I just tested and for a `bytes memory` it behaves the same way..

> - Function modifiers are implemented in a slightly different way regarding function parameters.

Reviewer 2: ```suggestion
- Function modifiers are implemented in a slightly different way regarding function parameters and return variables.
```

  This especially has an effect if the placeholder ``_;`` is evaluated multiple times in a modifier.
  In the old code generator, each function parameter and return parameter has a fixed slot on the stack.
  If the function is run multiple times because ``_;`` is used multiple times or used in a loop, then a

>   change to the function (return-)parameter's value is visible in the next execution of the function.

Reviewer 2: ```suggestion
  In the old code generator, each function parameter and return variable has a fixed slot on the stack.
  If the function is run multiple times because ``_;`` is used multiple times or used in a loop, then a
  change to the function parameter's or return variable's value is visible in the next execution of the function.
```

  The new code generator implements modifiers using actual functions and passes function parameters on.

>   This means that multiple executions of a function will get the same values for the parameters.

Reviewer 2: ```suggestion
  This means that multiple executions of a function's body will get the same values for the parameters,
  and the effect on return variables is that they are reset to their default (zero) value for each
  execution.
```

Author: I don't think a `,` before an `and` is correct?

Reviewer 2: It is, it just changes the meaning. We can also use a `;` if you prefer that or just use `. The effect...`

Author: I just read up. You're right. I have learned something :)

  If you execute ``f(0)`` in the old code generator, it will return ``2``, while
  it will return ``1`` when using the new code generator.

  .. code-block:: solidity

      // SPDX-License-Identifier: GPL-3.0
      pragma solidity >=0.7.1 <0.9.0;

      contract C {
          bool active;

>           constructor() { active = true; }

Reviewer 2: ```suggestion
          bool active = true;
```

          modifier mod()
          {
              _;
              active = false;
              _;
          }

>           function foo() external mod payable returns(uint ret)

Reviewer 2: ```suggestion
          function foo() external mod() returns (uint ret)
```

          {
              if (active)

>                   return 1;

Reviewer 2: Actually I wonder if this example is easier to understand if we just replace this line by `ret = 1`.
Can you check if it is still valid in that case? We can add a comment that `return 1` would be have in the same way.

Author: Yep, still valid. Changed it.

          }
      }

  The function ``C.foo()`` returns the following values:

>   - Old code generator: ``1`` as the return variable is initialized to ``0`` only once before the first call and then overwritten by the ``return 1;``. It's not initialized again for the second call and ``foo()`` doesn't explicitly assign it either (due to ``active == false``), thus it keeps its first value.

Reviewer 2: ```suggestion
  - Old code generator: ``1`` as the return variable is initialized to ``0`` only once before the first call and
    then overwritten by the ``return 1;``. It is not initialized again for the second call and ``foo()`` does not
    explicitly assign it either (due to ``active == false``), thus it keeps its first value.
```

>   - New code generator: ``0`` as all parameters, including return parameters, will be initialized again before each call.

Reviewer 2: Here, but also in the rest of the changes: I would avoid using the term `call` but instead use `function body execution` or `evaluation of the function body` or `placeholder run` or something like that. Otherwise it could be confused with the outermost call, which is the only real function call.

  If you execute ``f(0)`` in the old code generator, it will return ``2``, while
  it will return ``1`` when using the new code generator.

> - In some cases modifiers also behave differently in regards to the return

Reviewer 2: Can you try to combine parameters and return parameters? It's exactly the same effect.

  parameters.

  .. code-block:: solidity

      // SPDX-License-Identifier: GPL-3.0
      pragma solidity >=0.7.1 <0.9.0;

      contract C {
          bool active;
          constructor() { active = true; }
          modifier mod()
          {
              _;
              active = false;
              _;
          }
          function foo() external mod payable returns(uint ret)
          {
              if (active)
                  return 1;
          }
      }

  The function ``C.foo()`` returns the following values:

>   - Old code generator: ``1`` due to the return variable never getting assigned and thus it never overwrites the first return value in ``mod()``

Reviewer: Could you clarify what is meant by return variable being assigned and the distinction between return value and return variable?

As far as I understood, there are two placeholder statements in `mod` both of which invoke the body of `foo()`. The first invocation returns `1` but the return value (`1`) is not assigned. The second invocation in legacy (old) codegen returns `1` because the return value on the stack is not reset (?). However, IR resets it before the second invocation which is why it returns `0`.

Reviewer 2: The `return` statement in solidity is equivalent to an assignment to the return variables (even if they are not named) plus exiting the function. This means `_` does not return anything, it executes the function body, which assigns the return variables (which are not accessible directly in the modifier). In the new codegen, the second placeholder resets the values of the return variables while in the old codegen, they are only reset at the very beginning, not with each `_`.

Author: I don't think that's exactly correct. They are also reset with the second `_`, but only if we assign something to the return variable (e.g. explicitly `return 0;` at the end of the function)

Author: Nevermind, you basically said the same, I just didn't understand it correctly at first

  If you execute ``f(0)`` in the old code generator, it will return ``2``, while
  it will return ``1`` when using the new code generator.

  .. code-block:: solidity

      // SPDX-License-Identifier: GPL-3.0
      pragma solidity >=0.7.1 <0.9.0;

      contract C {
          bool active = true;
          modifier mod()
          {
              _;
              active = false;
              _;
          }
          function foo() external mod() returns (uint ret)
          {
              if (active)
                  ret = 1; // Same as ``return 1``
          }
      }

  The function ``C.foo()`` returns the following values:

  - Old code generator: ``1`` as the return variable is initialized to ``0`` only once before the first ``_;``
    evaluation and then overwritten by the ``return 1;``. It is not initialized again for the second ``_;``
    evaluation and ``foo()`` does not explicitly assign it either (due to ``active == false``), thus it keeps
    its first value.

>   - New code generator: ``0`` as all parameters, including return parameters, will be initialized again before

Reviewer: ```suggestion
  - New code generator: ``0`` as all parameters, including return parameters, will be re-initialized before
```

Note that, unlike the new code generator, the old code generator does not perform a cleanup after the bit-not assignment (``_a = ~_a``).
This results in different values being assigned (within the inline assembly block) to return value ``_r1`` between the old and new code generators.
However, both code generators perform a cleanup before the new value of ``_a`` is assigned to ``_r2``.

.. _modifier_codegen_differences:

> Modifiers

Reviewer 2: There is a section on modifiers already in line 38

Reviewer 2: It does not talk about return parameters, I think we should add that they are reset to their zero value for each `_`.
